core/services/offchainreporting/contract_tracker.go

```diff
@@ -0,0 +1,329 @@
+	if !t.OkayToStart() {
+		return errors.New("OCRContractTracker: already started")
+	}
+	connected := t.logBroadcaster.Register(t.contractAddress, t)
+	if !connected {
+		t.logger.Warnw("OCRContractTracker#Start: log broadcaster is not connected", "jobID", t.jobID, "address", t.contractAddress)
+	}
+	t.wg.Add(1)
+	go t.processLogs()
+	return nil
+}
+
+// Close should be called after teardown of the OCR job relying on this tracker
+func (t *OCRContractTracker) Close() error {
+	if !t.OkayToStop() {
+		return errors.New("OCRContractTracker already stopped")
+	}
+	close(t.chStop)
+	t.wg.Wait()
+	t.logBroadcaster.Unregister(t.contractAddress, t)
+	close(t.chConfigs)
+	return nil
+}
+
+func (t *OCRContractTracker) processLogs() {
+	defer t.wg.Done()
+	for {
+		select {
+		case <-t.configsMB.Notify():
+			// NOTE: libocr could take an arbitrary amount of time to process a
+			// new config. To avoid blocking the log broadcaster, we use this
+			// background thread to deliver them and a mailbox as the buffer.
+			for {
+				x := t.configsMB.Retrieve()
+				if x == nil {
+					break
+				}
+				cc, ok := x.(ocrtypes.ContractConfig)
+				if !ok {
+					panic(fmt.Sprintf("expected ocrtypes.ContractConfig but got %T", x))
+				}
+				select {
+				case t.chConfigs <- cc:
+				case <-t.chStop:
+					return
+				}
+			}
+		case <-t.chStop:
+			return
+		}
+	}
+}
+
+// OnConnect complies with LogListener interface
+func (t *OCRContractTracker) OnConnect() {}
+
+// OnDisconnect complies with LogListener interface
+func (t *OCRContractTracker) OnDisconnect() {}
+
+// HandleLog complies with LogListener interface
+func (t *OCRContractTracker) HandleLog(lb log.Broadcast, err error) {
+	if err != nil {
+		t.logger.Errorw("OCRContract: error in previous LogListener", "err", err)
+		return
+	}
+
+	was, err := lb.WasAlreadyConsumed()
+	if err != nil {
+		t.logger.Errorw("OCRContract: could not determine if log was already consumed", "error", err)
+		return
+	} else if was {
+		return
+	}
+
+	defer logger.ErrorIfCalling(lb.MarkConsumed)
+
+	topics := lb.RawLog().Topics
+	if len(topics) == 0 {
+		return
+	}
+	raw := lb.RawLog()
+	if raw.Address != t.contractAddress {
+		t.logger.Errorf("log address of 0x%x does not match configured contract address of 0x%x", raw.Address, t.contractAddress)
+		return
+	}
+	switch topics[0] {
+	case OCRContractConfigSet:
+		var configSet *offchainaggregator.OffchainAggregatorConfigSet
+		configSet, err = t.contractFilterer.ParseConfigSet(raw)
+		if err != nil {
+			t.logger.Errorw("could not parse config set", "err", err)
+			return
+		}
+		configSet.Raw = lb.RawLog()
+		cc := confighelper.ContractConfigFromConfigSetEvent(*configSet)
+
+		t.configsMB.Deliver(cc)
+	case OCRContractLatestRoundRequested:
+		var rr *offchainaggregator.OffchainAggregatorRoundRequested
+		rr, err = t.contractFilterer.ParseRoundRequested(raw)
+		if err != nil {
+			t.logger.Errorw("could not parse round requested", "err", err)
```

Comment:
Does the caller of this also log? Could these log statements result in multiple log outputs for a single error?

Reply:
Nah because `HandleLog` does not return error.